File: tarea-python-a-javascript/sample-functions.py
"""
Descripción:
Este archivo contiene 15 funciones básicas en Python para practicar lógica y programación.
El objetivo es que los alumnos las traduzcan a JavaScript, entendiendo los conceptos
de variables, concatenación, estructuras de control, arreglos (listas), y funciones.
"""

def saludar_usuario(nombre: str) -> None:
    """
    Imprime un saludo personalizado.
    Ejemplo de concatenación y salida por consola.
    """
    print("Hola, " + nombre + "! Bienvenido al curso de programación.")


def repetir_texto(texto: str, veces: int) -> str:
    """
    Devuelve una cadena repetida 'veces' veces, separada por espacios.
    Ejemplo de concatenación y multiplicación de strings.
    """
    return (texto + " ") * veces


def invertir_palabra(palabra: str) -> str:
    """
    Invierte una palabra usando slicing.
    """
    return palabra[::-1]


def contar_vocales(texto: str) -> int:
    """
    Cuenta cuántas vocales (a, e, i, o, u) hay en un texto.
    Ejemplo de bucles y condicionales.
    """
    contador = 0
    for letra in texto.lower():
        if letra in "aeiou":
            contador += 1
    return contador


def mayusculas_y_minusculas(texto: str) -> tuple[str, str]:
    """
    Devuelve el texto en mayúsculas y minúsculas.
    Ejemplo de métodos de string.
    """
    return texto.upper(), texto.lower()


def promedio_lista(numeros: list[float]) -> float:
    """
    Calcula el promedio de una lista de números.
    Ejemplo de sumas y longitud de listas.
    """
    if len(numeros) == 0:
        return 0
    return sum(numeros) / len(numeros)


def maximo_y_minimo(numeros: list[int]) -> tuple[int, int]:
    """
    Devuelve el valor máximo y mínimo de una lista.
    Ejemplo de uso de funciones integradas.
    """
    return max(numeros), min(numeros)


def filtrar_pares(numeros: list[int]) -> list[int]:
    """
    Filtra y devuelve solo los números pares de una lista.
    Ejemplo de comprensión de listas.
    """
    return [n for n in numeros if n % 2 == 0]


def sumar_elementos_texto(lista_textos: list[str]) -> str:
    """
    Une todos los elementos de una lista de cadenas en una sola frase.
    Ejemplo de concatenación con 'join'.
    """
    return " ".join(lista_textos)


def buscar_elemento(lista: list, elemento) -> bool:
    """
    Verifica si un elemento está dentro de una lista.
    Ejemplo de uso del operador 'in'.
    """
    return elemento in lista


def contar_palabras(frase: str) -> int:
    """
    Cuenta cuántas palabras hay en una frase.
    Usa split() para separar por espacios.
    """
    palabras = frase.split()
    return len(palabras)


def duplicar_elementos(lista: list[int]) -> list[int]:
    """
    Duplica cada número de una lista.
    Ejemplo de transformación con comprensión de listas.
    """
    return [x * 2 for x in lista]


def capitalizar_palabras(frase: str) -> str:
    """
    Convierte la primera letra de cada palabra en mayúscula.
    Similar a 'title()'.
    """
    return frase.title()


def mezclar_listas(lista1: list, lista2: list) -> list:
    """
    Combina dos listas alternando sus elementos.
    Ejemplo de zip() y bucles.
    """
    combinada = []
    for a, b in zip(lista1, lista2):
        combinada.append(a)
        combinada.append(b)
    return combinada


def contar_frecuencia(lista: list) -> dict:
    """
    Devuelve un diccionario con la frecuencia de cada elemento.
    Ejemplo de diccionarios y conteo.
    """
    conteo = {}
    for item in lista:
        if item in conteo:
            conteo[item] += 1
        else:
            conteo[item] = 1
    return conteo

File: tarea-python-a-javascript/sample-functions.js
/** Imprime saludo personalizado.
 *  Salida por consola.
 */
function saludarUsuario(nombre) {
  return console.log(
    "Hola " + nombre + "! Bievenido al curso de programación."
  );
}

/** Devielve una cadena repetida 'veces' veces, separada por espacios.*/
function repetirTexto(texto, veces) {
    return Array(veces).fill(texto).join(' ');
}

/** Comento la prueba de mi función */
// repetirTexto("Hola",3);

/** Explicación:
 *  1.- La función Array() crea un arreglo vacío.
 *  2.- Luego con .fill(texto) lleno el arreglo con ese texto en todos los elementos.
 *  3.- Finalmente con .join(separador) une todos los elementos con lo que pongas como separador.
*/ 

function invertirPalabra(palabra){
    return palabra.split('').reverse().join('');
}

/**
 * Explicación:
 * 1.- split('') hace que el string me lo convierta en un arreglo separado por cada caracter.
 * 2.- reverse() invierte el arreglo-
 * 3.- con join('') uno los eleemtnos del arreglo invertido y los junta.
 */

function contarVocales(texto) {
  let contador = 0;
  const vocales = "aeiou";
  for (let letra of texto.toLowerCase()) {
    if (vocales.includes(letra)) {
      contador++;
    }
  }
  return contador;
}

/**
 * Explicación:
 * 1.- Convierto el texto a minúsculas para simplificar la comparación.
 * 2.- Recorro cada letra usando un for..
 * 3.- Uso .includes() para verificar si la letra está en la cadena de vocales.
 * 4.- Si es vocal, incremento el contador.
 */

function mayusculasYMinusculas(texto) {
  return [texto.toUpperCase(), texto.toLowerCase()];
}

/**
 * Explicación:
 * 1.- .toUpperCase() convierte todo el texto a mayúsculas.
 * 2.- .toLowerCase() lo convierte a minúsculas.
 * 3.- Devuelvo ambos valores dentro de un arreglo.
 */

function promedioLista(numeros) {
  if (numeros.length === 0) return 0;
  const suma = numeros.reduce((acc, n) => acc + n, 0);
  return suma / numeros.length;
}

/**
 * Explicación:
 * 1.- Si la lista está vacía, retorno 0.
 * 2.- Uso .reduce() para sumar todos los elementos.
 * 3.- Divido entre la longitud del arreglo.
 */

function maximoYMinimo(numeros) {
  return [Math.max(...numeros), Math.min(...numeros)];
}

/**
 * Explicación:
 * 1.- Math.max() y Math.min() calculan el valor máximo y mínimo.
 * 2.- El operador spread (...) expande el arreglo para pasarlo como argumentos.
 */

function filtrarPares(numeros) {
  return numeros.filter((n) => n % 2 === 0);
}

/**
 * Explicación:
 * 1.- .filter() crea un nuevo arreglo con los elementos que cumplen la condición.
 * 2.- La condición es que el número sea divisible entre 2.
 */

function sumarElementosTexto(listaTextos) {
  return listaTextos.join(" ");
}

/**
 * Explicación:
 * 1.- .join(" ") une todos los elementos del arreglo separados por un espacio.
 */

function buscarElemento(lista, elemento) {
  return lista.includes(elemento);
}

/**
 * Explicación:
 * 1.- .includes() devuelve true si el elemento está en el arreglo.
 * 2.- Si no está, devuelve false.
 */

function contarPalabras(frase) {
  const palabras = frase.trim().split(/\s+/);
  return palabras.length;
}

/**
 * Explicación:
 * 1.- trim() elimina espacios al inicio y al final.
 * 2.- split(/\s+/) divide por uno o más espacios.
 * 3.- Devuelvo la cantidad de palabras usando .length.
 */

function duplicarElementos(lista) {
  return lista.map((x) => x * 2);
}

/**
 * Explicación:
 * 1.- .map() transforma cada elemento según la función que se le pasa.
 * 2.- Aquí multiplico cada valor por 2.
 */

function capitalizarPalabras(frase) {
  return frase
    .split(" ")
    .map((palabra) => palabra.charAt(0).toUpperCase() + palabra.slice(1))
    .join(" ");
}

/**
 * Explicación:
 * 1.- split(" ") separa la frase en palabras.
 * 2.- Para cada palabra, convierto la primera letra a mayúscula.
 * 3.- slice(1) conserva el resto.
 * 4.- join(" ") une todo de nuevo.
 */

function mezclarListas(lista1, lista2) {
  const combinada = [];
  const longitud = Math.min(lista1.length, lista2.length);
  for (let i = 0; i < longitud; i++) {
    combinada.push(lista1[i], lista2[i]);
  }
  return combinada;
}

/**
 * Explicación:
 * 1.- Uso un bucle for con el índice más corto de las dos listas.
 * 2.- En cada iteración agrego un elemento de cada lista.
 */

function contarFrecuencia(lista) {
  const conteo = {};
  for (let item of lista) {
    if (conteo[item]) {
      conteo[item]++;
    } else {
      conteo[item] = 1;
    }
  }
  return conteo;
}

/**
 * Explicación:
 * 1.- Creo un objeto vacío.
 * 2.- Recorro la lista y verifico si el elemento ya existe como clave.
 * 3.- Si existe, incremento el contador, y ps si no, lo inicializo en 1.
 */

